### PR TITLE
Enables the prescription of a distribution mean from a previous HM

### LINKED ISSUE #66: Use posterior ensemble mean for new prior
Practical experience shows that history matching results may still contain biases. A two-step solution approach was proposed by Guo and Reynolds (2019) to address this, in which the posterior ensemble means are used as the means for a new prior in a second application of the history matching workflow.

If the ensemble of parameter estimates is saved in a compact form at the end of the history matching, new functionality could be introduced to specify that the mean of the new prior should be taken from this file.

### PR DESCRIPTION
**Pull request by olwijn**
Closes #66 

The functionality is activated by adding the keyword ahm_case to the model_parameters section of the config:

image

The specified folder is currently expected to contain the file parameters_iteration-latest.parquet.gzip

### CODE REVIEW

File: src/flownet/ahm/_run_ahm.py [hunk 1]
```diff
@@ -119,6 +122,94 @@ def _get_distribution(
...
+        n = len(parameter.random_variables)
+
+        for i, var in enumerate(parameter.random_variables):
+            mean = parameter.mean_values[i]
+
+            loguniform = var.ert_gen_kw.split()[0] == "LOGUNIF"
+            dist_min0 = var.minimum
+            dist_max0 = var.maximum
+
+            if loguniform is True:
```
[the remark this answers is not in the document]
Comment (olwijn) [reply src/flownet/ahm/_run_ahm.py:207]: I have now created a separate function to determine min/max

File: src/flownet/ahm/_run_ahm.py [hunk 2]
```diff
@@ -119,6 +122,94 @@ def _get_distribution(
     return df
 
 
+def update_distribution(parameters, ahm_case):
```
[the remark this answers is not in the document]
Comment (olwijn) [reply src/flownet/ahm/_run_ahm.py]: type annotation was added

**Comment by olwijn** [comment]
image

I don't quite understand the problem. reveal_type(parameters) on input and output gives:

image

**Comment by wouterjdb** [comment]
Mypy is complaining about that you tried to get attributes from something which is allowed to be a str, according to the input type,

`parameters: Union[str, List[str]]`

and a string does not have these attributes.

**Comment by olwijn** [comment]
Two more changes are needed to fully integrate this functionality: (1) add a test to workflow.yml in .github/workflows : add a second history match run with restart after the first one and before the prediction run, (2) do not rebuild the flownet when using the restart, but instead use the functionality used for prediction runs to read in the flownet.

**Comment by olwijn** [comment]
Two more changes are needed to fully integrate this functionality: (1) add a test to workflow.yml in .github/workflows : add a second history match run with restart after the first one and before the prediction run, (2) do not rebuild the flownet when using the restart, but instead use the functionality used for prediction runs to read in the flownet.

File: .github/workflows/flownet.yml [hunk 3]
```diff
@@ -88,6 +88,7 @@ jobs:
           flownet ahm ./config/assisted_history_matching.yml ./some_ahm_run --update-config ./ci_config/assisted_history_matching.yml
```
Comment (anders-kiaer) [inline .github/workflows/flownet.yml:88]: Can we add one line here where the new functionality is used? Then we are sure the added functionality runs :green_heart: :slightly_smiling_face:

File: src/flownet/_command_line.py [hunk 4]
```diff
@@ -144,6 +144,13 @@ def main():
         help="Optional configuration file which values will update the main config. "
         "Any relative paths in this file will also be assumed relative to the main config.",
     )
+    parser_ahm.add_argument(
+        "--restart-folder",
+        type=pathlib.Path,
+        default=None,
+        help="Optional folder containing results from a previous history match. "
+        "Any relative paths in this file will also be assumed relative to the main config.",
```
Comment (wouterjdb) [inline src/flownet/_command_line.py:152]:
```suggestion
        help="Optional folder containing results from a previous history match.",
```

File: src/flownet/ahm/_assisted_history_matching.py [hunk 5]
```diff
@@ -94,7 +94,7 @@ def run_ert(self, weights: List[float]):
         Args:
             weights: Weights for the iterated ensemble smoother to use.
 
-        Returns:
+        Returns:self.output_folder
```
Comment (wouterjdb) [inline src/flownet/ahm/_assisted_history_matching.py:97]:
```suggestion
        Returns:
```

**Review by wouterjdb — CHANGES_REQUESTED** [review]
LGTM - two minor things. You can change those from within your browser (accept the suggestions). Approved afterward. 👍

File: src/flownet/ahm/_run_ahm.py [hunk 6]
```diff
@@ -1,6 +1,10 @@
 import argparse
 from typing import Union, List, Optional
```
Comment (anders-kiaer) [inline src/flownet/ahm/_run_ahm.py:3]:
```suggestion

```

**Review by anders-kiaer — COMMENTED** [review]
One minor suggestion, otherwise LGTM as well.

**Review by wouterjdb — APPROVED** [review]

**Pull request merged**